service/frontend/adminHandler.go

```diff
@@ -0,0 +1,121 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/admin"
+	"github.com/uber/cadence/.gen/go/admin/adminserviceserver"
+	gen "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/service"
+)
+
+var _ adminserviceserver.Interface = (*AdminHandler)(nil)
+
+type (
+	// AdminHandler - Thrift handler inteface for admin service
+	AdminHandler struct {
+		numberOfHistoryShards int
+		service.Service
+	}
+)
+
+// NewAdminHandler creates a thrift handler for the cadence admin service
+func NewAdminHandler(
+	sVice service.Service, numberOfHistoryShards int) *AdminHandler {
+	handler := &AdminHandler{
+		numberOfHistoryShards: numberOfHistoryShards,
+		Service:               sVice,
+	}
+	return handler
+}
+
+// Start starts the handler
+func (adh *AdminHandler) Start() error {
+	adh.Service.GetDispatcher().Register(adminserviceserver.New(adh))
+	adh.Service.Start()
+	return nil
+}
+
+// Stop stops the handler
+func (adh *AdminHandler) Stop() {
+	adh.Service.Stop()
+}
+
+// InquiryWorkflowExecution returns information about the specified workflow execution.
+func (adh *AdminHandler) InquiryWorkflowExecution(ctx context.Context, request *gen.DescribeWorkflowExecutionRequest) (*admin.InquiryWorkflowExecutionResponse, error) {
+	if request == nil {
+		return nil, adh.error(errRequestNotSet)
+	}
+
+	if err := adh.validateExecution(request.Execution); err != nil {
+		return nil, err
+	}
+
+	shardID := common.WorkflowIDToHistoryShard(*request.Execution.WorkflowId, adh.numberOfHistoryShards)
+	shardIDstr := string(shardID)
+
+	historyHost, err := adh.GetMembershipMonitor().Lookup(common.HistoryServiceName, shardIDstr)
+	historyAddr := historyHost.GetAddress()
+
+	if err != nil {
+		return nil, adh.error(err)
+	}
+
+	return &admin.InquiryWorkflowExecutionResponse{
+		ShardId:     &shardIDstr,
+		HistoryAddr: &historyAddr,
+	}, nil
+}
+
+func (adh *AdminHandler) error(err error) error {
+	switch err.(type) {
+	case *gen.InternalServiceError:
+		logging.LogInternalServiceError(adh.Service.GetLogger(), err)
+		return err
+	case *gen.BadRequestError:
+		return err
+	case *gen.ServiceBusyError:
+		return err
+	case *gen.EntityNotExistsError:
+		return err
+	default:
+		logging.LogUncategorizedError(adh.Service.GetLogger(), err)
+		return &gen.InternalServiceError{Message: err.Error()}
+	}
+}
+
+func (adh *AdminHandler) validateExecution(w *gen.WorkflowExecution) error {
```

Comment:
I wanted to do that at the beginning. It is going to refactor error, and also need to define all metrics(about 10) for a new scope for admin service. But I felt like these will be too heavy for admin API at this point. It will introduce more code, and emit lots of metrics that we are not going to use.

Reply:
something from `func (wh *WorkflowHandler) validateExecution(w *gen.WorkflowExecution, scope int) error ` to `func validateExecution(w *gen.WorkflowExecution, scope int) error '; from `func (wh *WorkflowHandler) error(err error, scope int) error` to `func error(err error, scope int, logger, metric...) error`